translator/internaldata/metrics_to_oc.go

```diff
@@ -0,0 +1,362 @@
+						SumOfSquaredDeviation: 0,
+						BucketOptions:         histogramExplicitBoundsToOC(point.ExplicitBounds()),
+						Buckets:               histogramBucketsToOC(point.Buckets()),
+					},
+				},
+			},
+		},
+	}
+}
+
+func histogramExplicitBoundsToOC(bounds []float64) *ocmetrics.DistributionValue_BucketOptions {
+	if len(bounds) == 0 {
+		return nil
+	}
+
+	return &ocmetrics.DistributionValue_BucketOptions{
+		Type: &ocmetrics.DistributionValue_BucketOptions_Explicit_{
+			Explicit: &ocmetrics.DistributionValue_BucketOptions_Explicit{
+				Bounds: bounds,
+			},
+		},
+	}
+}
+
+func histogramBucketsToOC(buckets []data.HistogramBucket) []*ocmetrics.DistributionValue_Bucket {
+	if len(buckets) == 0 {
+		return nil
+	}
+
+	ocBuckets := make([]*ocmetrics.DistributionValue_Bucket, 0, len(buckets))
+	for _, bucket := range buckets {
+		ocBuckets = append(ocBuckets, &ocmetrics.DistributionValue_Bucket{
+			Count:    int64(bucket.Count()),
+			Exemplar: exemplarToOC(bucket.Exemplar()),
+		})
+	}
+	return ocBuckets
+}
+
+func exemplarToOC(exemplar data.HistogramBucketExemplar) *ocmetrics.DistributionValue_Exemplar {
+	return &ocmetrics.DistributionValue_Exemplar{
+		Value:       exemplar.Value(),
+		Timestamp:   internal.UnixnanoToTimestamp(exemplar.Timestamp()),
+		Attachments: exemplar.Attachments(),
+	}
+}
+
+func summaryPointToOC(point data.SummaryDataPoint, labelKeys *labelKeys) *ocmetrics.TimeSeries {
+	return &ocmetrics.TimeSeries{
+		StartTimestamp: internal.UnixnanoToTimestamp(point.StartTime()),
+		LabelValues:    labelValuesToOC(point.LabelsMap(), labelKeys),
+		Points: []*ocmetrics.Point{
+			{
+				Timestamp: internal.UnixnanoToTimestamp(point.Timestamp()),
+				Value: &ocmetrics.Point_SummaryValue{
+					SummaryValue: &ocmetrics.SummaryValue{
+						Count: int64Value(point.Count()),
+						Sum:   doubleValue(point.Sum()),
+						Snapshot: &ocmetrics.SummaryValue_Snapshot{
+							Count:            nil,
+							Sum:              nil,
+							PercentileValues: percentileToOC(point.ValueAtPercentiles()),
+						},
+					},
+				},
+			},
+		},
+	}
+}
+
+func percentileToOC(percentiles []data.SummaryValueAtPercentile) []*ocmetrics.SummaryValue_Snapshot_ValueAtPercentile {
+	if len(percentiles) == 0 {
+		return nil
+	}
+
+	ocPercentiles := make([]*ocmetrics.SummaryValue_Snapshot_ValueAtPercentile, 0, len(percentiles))
+	for _, p := range percentiles {
+		ocPercentiles = append(ocPercentiles, &ocmetrics.SummaryValue_Snapshot_ValueAtPercentile{
+			Percentile: p.Percentile(),
+			Value:      p.Value(),
+		})
+	}
+	return ocPercentiles
+}
+
+func labelValuesToOC(labelsMap data.LabelsMap, labelKeys *labelKeys) []*ocmetrics.LabelValue {
+	if len(labelsMap) == 0 {
+		return nil
+	}
+
+	// Initialize label values with defaults
+	// (The order matches key indices)
+	labelValues := make([]*ocmetrics.LabelValue, len(labelKeys.keyIndices))
+	for i := 0; i < len(labelKeys.keys); i++ {
+		labelValues[i] = &ocmetrics.LabelValue{
+			HasValue: false,
```

Comment:
not from your change but would have been better if OC treated nil to mean no value... side note: `HasValue: false,` is not strictly necessary.